### PR TITLE
Error handling for CourseSchedule VC

### PR DESCRIPTION
**Pull request by 01jongmin**
Changed return type for `getCourses` function of `PennInTouchNetworkManager` to `Result`

Added error handling for `courseSchedule` View Controller

Videos:
[Refresh Controller](https://youtu.be/MYdAYmIyiGQ) 
[Login Logout remove course schedule](https://youtu.be/D4r11miZ-KQ)

### CODE REVIEW

File: PennMobile/Courses/CourseScheduleViewController.swift [hunk 1]
```diff
@@ -56,18 +34,143 @@ extension CourseScheduleViewController {
...
+            }
+        })
+    }
+
+    // TODO: Need review: ad-hoc method to overcome conflict of UIAlertController and UIRefreshController
+    override func viewDidDisappear(_ animated: Bool) {
+        super.viewDidDisappear(animated)
+        self.courseScheduleTableView.isHidden = true
+        self.courseScheduleTableView.refreshControl?.endRefreshing()
+        self.courseScheduleTableView.contentOffset = CGPoint.zero
```
Comment (01jongmin) [inline PennMobile/Courses/CourseScheduleViewController.swift]: This is an ad-hoc method.
For some reason, UIAlertController and UIRefreshController seems to be in conflict with one another that prevents me from seeing the effect of calling `endRefreshing()`

Similar issues has been brought up by other [people](https://stackoverflow.com/questions/40230892/uirefresh-control-endrefreshing-doesnt-work). I wasn't sure if you knew of a better work around way of doing this.

File: PennMobile/Courses/CourseScheduleViewController.swift [hunk 2]
```diff
@@ -56,18 +34,143 @@ extension CourseScheduleViewController {
...
+            self.showAlert(withMsg: "You appear to be offline.\nPlease try again later.", title: "Network Error", completion: { self.navigationController?.popViewController(animated: true) } )
+
+        case .failure(.parsingError):
+            self.showAlert(withMsg: "Penn's course servers are currently not updating. We hope this will be fixed shortly", title: "Uh oh!", completion: { self.navigationController?.popViewController(animated: true) })
+
+        case .failure(.authenticationError):
+            self.showAlert(withMsg: "Unable to access your courses.\nPlease login again.", title: "Login Error", completion: { self.handleAuthentication() })
+
+        default:
+            self.showAlert(withMsg: "Something went wrong.\nPlease try again later.", title: "Uh oh!", completion: { self.courseScheduleTableView.refreshControl?.endRefreshing(); self.navigationController?.popViewController(animated: true) } )
```
Comment (joshdoman) [inline PennMobile/Courses/CourseScheduleViewController.swift]: Let's look into abstracting this error handling into a protocol, perhaps one that inherits from ShowsAlert (and thus must be a view controller). We should avoid duplicate code, especially when the error messages are the same.

Comment (01jongmin) [reply PennMobile/Courses/CourseScheduleViewController.swift]: Got it. I'll work on it right away!

File: PennMobile/General/Protocols.swift [hunk 3]
```diff
@@ -121,6 +121,37 @@ extension ShowsAlert where Self: UIViewController {
...
+            self.showAlert(withMsg: "You appear to be offline.\nPlease try again later.", title: "Network Error", completion: noInternet)
+
+        case .failure(.parsingError):
+            self.showAlert(withMsg: "Penn's \(title) servers are currently not updating. We hope this will be fixed shortly", title: "Uh oh!", completion: parsingError)
+            
+        case .failure(.serverError):
+            self.showAlert(withMsg: "Penn's \(title) servers are currently not updating. We hope this will be fixed shortly", title: "Uh oh!", completion: serverError)
+            
+        case .failure(.jsonError):
+            self.showAlert(withMsg: "Penn's \(title) servers are currently not updating. We hope this will be fixed shortly", title: "Uh oh!", completion: jsonError)
```
Comment (01jongmin) [inline PennMobile/General/Protocols.swift]: I wasn't too sure if it was the correct decision to make the messaging for internet, parsing, and json error all the same, informing the users that the issue wasn't on our end.

[the remark this answers is not in the document]
Comment (r3khaan) [reply PennMobile/General/Protocols.swift]: Now that I think about it, there's also a strong case for just changing the name. Adding extra functionality for customizing the success message & title is something that would only be useful later (when we have more than 1 type of success message). We can always change it then.

[the remark this answers is not in the document]
Comment (01jongmin) [reply PennMobile/General/Protocols.swift]: What about we keep the protocol name as `ShowsAlertForError` and change the function name to `showRefreshAlertForError`. We can then add additional functions that respond to error for different type of actions later down the line.

**Pull request merged**